File: src/main/java/dev/slne/protect/bukkit/regions/QuickHull.java
package dev.slne.protect.bukkit.regions;

import java.util.ArrayList;
import java.util.List;

import dev.slne.protect.bukkit.visual.Marker;

public class QuickHull {

	public List<Marker> quickHull(List<Marker> points) {
		ArrayList<Marker> convexHull = new ArrayList<Marker>();
		if (points.size() < 3)
			return points;

		int minMarker = -1, maxMarker = -1;
		int minX = Integer.MAX_VALUE;
		int maxX = Integer.MIN_VALUE;
		for (int i = 0; i < points.size(); i++) {
			if (points.get(i).getX() < minX) {
				minX = points.get(i).getX();
				minMarker = i;
			}
			if (points.get(i).getX() > maxX) {
				maxX = points.get(i).getX();
				maxMarker = i;
			}
		}
		Marker A = points.get(minMarker);
		Marker B = points.get(maxMarker);
		convexHull.add(A);
		convexHull.add(B);
		points.remove(A);
		points.remove(B);

		ArrayList<Marker> leftSet = new ArrayList<Marker>();
		ArrayList<Marker> rightSet = new ArrayList<Marker>();

		for (int i = 0; i < points.size(); i++) {
			Marker p = points.get(i);
			if (pointLocation(A, B, p) == -1)
				leftSet.add(p);
			else if (pointLocation(A, B, p) == 1)
				rightSet.add(p);
		}
		hullSet(A, B, rightSet, convexHull);
		hullSet(B, A, leftSet, convexHull);

		return convexHull;
	}

	public int distance(Marker A, Marker B, Marker C) {
		int ABx = B.getX() - A.getX();
		int ABy = B.getZ() - A.getZ();
		int num = ABx * (A.getZ() - C.getZ()) - ABy * (A.getX() - C.getX());
		if (num < 0)
			num = -num;
		return num;
	}

	public void hullSet(Marker A, Marker B, ArrayList<Marker> set, ArrayList<Marker> hull) {
		int insertPosition = hull.indexOf(B);
		if (set.size() == 0)
			return;
		if (set.size() == 1) {
			Marker p = set.get(0);
			set.remove(p);
			hull.add(insertPosition, p);
			return;
		}
		int dist = Integer.MIN_VALUE;
		int furthestMarker = -1;
		for (int i = 0; i < set.size(); i++) {
			Marker p = set.get(i);
			int distance = distance(A, B, p);
			if (distance > dist) {
				dist = distance;
				furthestMarker = i;
			}
		}
		Marker P = set.get(furthestMarker);
		set.remove(furthestMarker);
		hull.add(insertPosition, P);

		// Determine who's to the left of AP
		ArrayList<Marker> leftSetAP = new ArrayList<Marker>();
		for (int i = 0; i < set.size(); i++) {
			Marker M = set.get(i);
			if (pointLocation(A, P, M) == 1) {
				leftSetAP.add(M);
			}
		}

		// Determine who's to the left of PB
		ArrayList<Marker> leftSetPB = new ArrayList<Marker>();
		for (int i = 0; i < set.size(); i++) {
			Marker M = set.get(i);
			if (pointLocation(P, B, M) == 1) {
				leftSetPB.add(M);
			}
		}
		hullSet(A, P, leftSetAP, hull);
		hullSet(P, B, leftSetPB, hull);

	}

	public int pointLocation(Marker A, Marker B, Marker P) {
		int cp1 = (B.getX() - A.getX()) * (P.getZ() - A.getZ()) - (B.getZ() - A.getZ()) * (P.getX() - A.getX());
		if (cp1 > 0)
			return 1;
		else if (cp1 == 0)
			return 0;
		else
			return -1;
	}
}

File: src/main/kotlin/dev/slne/surf/protect/paper/region/visual/QuickHull.kt
package dev.slne.surf.protect.paper.region.visual

import dev.slne.surf.surfapi.core.api.util.mutableObjectListOf
import dev.slne.surf.surfapi.core.api.util.objectListOf
import it.unimi.dsi.fastutil.objects.ObjectArrayList
import it.unimi.dsi.fastutil.objects.ObjectList
import kotlin.math.abs

/**
 * A branch‑lean, allocation‑light Quick‑Hull implementation that works on an
 * [ObjectList] of [Marker]s and returns the convex hull in counter‑clockwise
 * order. Uses *fastutil* collections to avoid boxing and minimise GC load.
 *
 * Time‑complexity: **O(n log n)** worst‑case, **O(n)** for random point sets.
 */
object QuickHull {
    /**
     * Computes the convex hull of the given marker list.
     *
     * @param input A list containing **at least three** distinct markers.
     * @return A new [ObjectList] that represents the convex hull. The caller
     *         owns the list; modifications will not affect the original input.
     */
    fun compute(input: ObjectArrayList<Marker>): ObjectArrayList<Marker> {
        if (input.size < 3) return input

        val points = mutableObjectListOf(input)
        val hull = mutableObjectListOf<Marker>()


        var minX = Int.MAX_VALUE
        var maxX = Int.MIN_VALUE
        var minIdx = -1
        var maxIdx = -1

        for (i in 0 until input.size) {
            val point = input[i]
            val x = point.blockX
            if (x < minX) {
                minX = x
                minIdx = i
            }

            if (x > maxX) {
                maxX = x
                maxIdx = i
            }
        }

        val a = points[minIdx]
        val b = points[maxIdx]

        val hullPoints = objectListOf(a, b)
        hull.addAll(hullPoints)
        points.removeAll(hullPoints)

        val left = mutableObjectListOf<Marker>()
        val right = mutableObjectListOf<Marker>()

        for (point in points) {
            when (pointSide(a, b, point)) {
                -1 -> left.add(point)
                1 -> right.add(point)
            }
        }

        for (i in input.indices) {
            val marker = input[i]

            if (pointSide(a, b, marker) == -1) {
                left.add(marker)
            } else if (pointSide(a, b, marker) == 1) {
                right.add(marker)
            }
        }

        buildHull(a, b, right, hull)
        buildHull(b, a, left, hull)

        return hull
    }


    fun buildHull(
        a: Marker,
        b: Marker,
        set: ObjectArrayList<Marker>,
        hull: ObjectList<Marker>
    ) {
        val insertPosition = hull.indexOf(b)
        if (set.isEmpty()) return

        if (set.size == 1) {
            val point = set.pop()
            hull.add(insertPosition, point)
            return
        }

        var maxDist = Int.MIN_VALUE
        var idx = -1
        for (i in 0 until set.size) {
            val point = set[i]
            val distance = distance(a, b, point)

            if (distance > maxDist) {
                maxDist = distance
                idx = i
            }
        }

        val point = set[idx]
        set.removeAt(idx)
        hull.add(insertPosition, point)

        // Partition remaining points into two sets: left of AP and PB
        val leftAP = mutableObjectListOf<Marker>()
        val leftPB = mutableObjectListOf<Marker>()
        for (q in set) {
            when {
                pointSide(a, point, q) == 1 -> leftAP.add(q)
                pointSide(point, b, q) == 1 -> leftPB.add(q)
            }
        }

        buildHull(a, point, leftAP, hull)
        buildHull(point, b, leftPB, hull)
    }

    @Suppress("NOTHING_TO_INLINE")
    private inline fun distance(a: Marker, b: Marker, c: Marker): Int {
        val abx = b.blockX - a.blockX
        val abz = b.blockZ - a.blockZ
        val area = abx * (a.blockZ - c.blockZ) - abz * (a.blockX - c.blockX)
        return abs(area)
    }

    @Suppress("NOTHING_TO_INLINE")
    private inline fun pointSide(a: Marker, b: Marker, c: Marker): Int {
        val value = (b.blockX - a.blockX) * (c.blockZ - a.blockZ) -
                (b.blockZ - a.blockZ) * (c.blockX - a.blockX)

        return when {
            value > 0 -> 1
            value < 0 -> -1
            else -> 0
        }
    }
}
